### ISSUE TITLE
Example model(s) for BusinessSimulation library will not simulate

### ISSUE THREAD

**Issue by gwr69** [OpenModelica/OpenModelica#12145]
### Description

The model [`BusinessSimulation.Examples.SimpleProductionChainII`](https://github.com/bslMS/BusinessSimulation/blob/release/BusinessSimulation/Examples/SimpleProductionChainII.mo) from the `release` branch of the [Business Simulation Library (BSL)](https://bsl-support.de/BusinessSimulation/BusinessSimulation.html) will not simulate in Open Modelica. Given that the model validates ok in OMEdit and Wolfram System Modeler v14, with the model also running fine in System Modeler, I wonder whether this may not in fact be a bug in OMEdit?

### Steps to Reproduce

Open [`BusinessSimulation.Examples.SimpleProductionChainII`](https://github.com/bslMS/BusinessSimulation/blob/release/BusinessSimulation/Examples/SimpleProductionChainII.mo) in OMEdit and select `Simulation>Simulate` from the menu.

### Expected Behavior

The model should simulate without any issues.

### Screenshots

I am getting a `Translation Error`:

```
[3] 11:46:14 Translation Error
[C:/dev/OM64bit/OMCompiler/Compiler/SimCode/SimCodeUtil.mo: 804:5-804:146]: Internal error function createSimCode failed [Transformation from optimised DAE to simulation code structure failed]
```

### Version and OS

 - OpenModelica Version: `v1.23.0-dev-381-g47c3c7528c (64-bit)`
 - OS: `Windows 10, 64 bit`
 - `MSL 4.0`

### Additional Context

This issue is part of the effort to minimize frictions between different tools running a Modelica library. See [this discuusion](https://github.com/OpenModelica/OpenModelica/discussions/12073).

**Comment by gwr69** [comment]
I note that using `Simulation>Check Model` will result in 12 out of 12 models from `BusinessSimulation.Examples` package completing "sucessfully."

5 out of 12 models will not simulate:

- `SimpleProductionChain.mo`
- `SimpleProductionChainII.mo`
- `SimpleProductionChainIII.mo`
- `SIR.mo`
- `ManagingEmployment.mo`

Using `Simulation>Check All Models` for `SimpleProductionChainII.mo` will give warnings that I honestly do not understand:

```
Error Buffer: 
[D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:10:5-10:48:writable] Warning: Connector inventory is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
[D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:11:5-11:58:writable] Warning: Connector installedBase is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
[D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:12:5-12:46:writable] Warning: Connector producing is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
[D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:13:5-13:56:writable] Warning: Connector shipping is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
[D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:14:5-14:71:writable] Warning: Connector scrapping is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
```

**Comment by perost** [comment]
> Using `Simulation>Check All Models` for `SimpleProductionChainII.mo` will give warnings that I honestly do not understand:
> 
> ```
> Error Buffer: 
> [D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:10:5-10:48:writable] Warning: Connector inventory is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
> [D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:11:5-11:58:writable] Warning: Connector installedBase is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
> [D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:12:5-12:46:writable] Warning: Connector producing is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
> [D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:13:5-13:56:writable] Warning: Connector shipping is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
> [D:/git-local/BusinessSimulation/BusinessSimulation/Examples/SimpleProductionChainII.mo:14:5-14:71:writable] Warning: Connector scrapping is not balanced: The number of potential variables (1) is not equal to the number of flow variables (0). 
> ```

Fixed in #12146. It won't solve any of the other issues you have, it was just an issue because `Check All Models` would check `SimpleProductionChainII.ModelOutput`, and we didn't handle checking an expandable connector class by itself properly.

**Comment by gwr69** [comment]
@perost What do the Translation Errors signify for the `SimpleProductionChainII` and the other models?

**Comment by perost** [comment]
> @perost What do the Translation Errors signify for the `SimpleProductionChainII` and the other models?

I'm not sure, it seems something goes wrong in the backend. Maybe @kabdelhak or @phannebohm has some clue?

**Comment by casella** [comment]
The problem is replicated by this MWE:
```
model TestArrayEquation
  parameter Integer N = 3;
  Real x[N];
  parameter Real init = 0;
equation
  der(x) = -x + ones(N);
initial equation
  x[1:N] = array(init/N for i in 1:3);
end TestArrayEquation;
```
Apparently, the old backend doesn't like slicing in array initial equations.

@gwr69, the old (i.e., current) backend has separate code to deal with regular equations and initial equations, which were added later to the language. The new backend has a unified architecture, and in fact the MWE above works fine with `--newBackend`.

We are only fixing issues with the old backend if it is really necessary to do so, since we plan to retire it in a couple of years when the new backend will be mature enough.

It turns out, there is no need to use slicing in [your code](https://github.com/bslMS/BusinessSimulation/blob/de64a3fac3e632c304ff4db589e8b60b892595a2/BusinessSimulation/Stocks/DelayN.mo#L25). PR bslMS/BusinessSimulation#5 should fix the issue.

**Comment by gwr69** [comment]
```
model TestArrayEquation
  parameter Integer N = 3;
  Real x[N];
  parameter Real init = 0;
equation
  der(x) = -x + ones(N);
initial equation
  x = array(init/N for i in 1:3); // or: fill( ... )
end TestArrayEquation;
```

should work, or not?

**Comment by gwr69** [comment]
Ok, that was a very easy fix for me to do and all models are not simulating in OMEdit.
Great!

**Comment by casella** [comment]
All's well that ends well. Note that there is no need to change the RHS of those initial equations, the only thing that currently doesn't work is the use of slicing in initial equation.

**Comment by casella** [comment]
Unfortunately the [test job](https://libraries.openmodelica.org/branches/master/BusinessSimulation_dev/BusinessSimulation_dev.html) still uses an an older version of the library, so you still can't see all green there. We're working on it.

**Comment by casella** [comment]
@gwr69 please let me know when you make a new release of BusinessSimulation, so I can change the support level in the package manager from experimental to fullSupport.

**Comment by gwr69** [comment]
> @gwr69 please let me know when you make a new release of BusinessSimulation, so I can change the support level in the package manager from experimental to fullSupport.

@casella I just released `v2.2.0` targeted at OMEdit: https://github.com/bslMS/BusinessSimulation

- all examples will run as of `v.1.23.0-dev` (April 3, 2024) and have been (reluctantly) set to run in seconds
- top level parameters have been moved to a component `theta` at top level scope for better user experience (for now)

**Comment by casella** [comment]
Very good!

In a couple of days when the pipeline is updated you will be able to access two reports [here](https://libraries.openmodelica.org/branches/master/): 
- [BusinessSimulation_dev](https://libraries.openmodelica.org/branches/master/BusinessSimulation_dev/BusinessSimulation_dev.html) runs on the main branch, so it can be used to check if you break something during development
- [BusinessSimulation](https://libraries.openmodelica.org/branches/master/BusinessSimulation/BusinessSimulation.html) will test the latest released version, currently 2.2.0. This can be used to check for regressions due to bugs introduced in OMC.

Regression test reports are issued daily [here](https://libraries.openmodelica.org/branches/history/master/) for the whole library testsuite, I keep them monitored, so I will see if something goes badly wrong on our side with your library.

As a last remark, OpenModelica has seamless support for MSL 3.2.3. Nevertheless, I would suggest you to convert it to MSL 4.0.0 at your earliest covenience, if you want third parties to use it. I'm not sure about SystemModeller, but you can for sure use OMC to do that.

**Comment by gwr69** [comment]
I do not follow, the current version is dependent on MSL 4.0?

**Comment by casella** [comment]
> I do not follow, the current version is dependent on MSL 4.0?

No it isn't. My suggestion is to make a new 3.0.0 version of your library depending on MSL 4.0.0. MSL 3.2.3 is starting to get seriously obsolete.

**Comment by gwr69** [comment]
It should be, then it is a bug. Where is it not compliant with MSL 4.0?

**Comment by casella** [comment]
Sorry, I missed your latest changes, all fine.

**Comment by casella** [comment]
The pipeline went through: https://libraries.openmodelica.org/branches/history/master/2024-04-04%2011:25:03..2024-04-05%2000:46:52.html

All models [work fine now](https://libraries.openmodelica.org/branches/master/BusinessSimulation/BusinessSimulation.html), except [BusinessSimulation.Converters.DiscreteDelay.DelayInformation](https://libraries.openmodelica.org/branches/master/BusinessSimulation/files/BusinessSimulation_BusinessSimulation.Converters.DiscreteDelay.DelayInformation.err), failing with
```
BusinessSimulation 2.2.0/Converters/DiscreteDelay/DelayInformation.mo:18:3-18:36:writable] Error: 
The model can't be instantiated due to top-level outer element 'modelSettings',
it may only be used as part of a simulation model.
```
I guess this is just a component model that got an experiment annotation by mistake, you should probably remove it.

Success! 😄

**Comment by casella** [comment]
BTW, did you check the results?

**Comment by gwr69** [comment]
I have not checked any results yet in depth, but what I saw quickly from running all examples looked good. You got to provide users with realistic chances to find bugs. :)

I had removed that old annotation in the development version for the `-wsm` build and then forgotten about it. That's what you get from having multiple builds...

Currently, I am working on Business Simulation Pro, which will go for synchronous time. Unfortunately, at the moment functions with arrays cause problems—no more `algorithm` sections—and I think I need to check this with Open Modelica too...

I will need to find a good way to run OMEdit on my Mac, though. Switching between two computers so I can have access to Windows is not good.
